Feedback result set

dcsiebe opened the pull request:
To store multiple audio files per feedback recording the following changes are necessary.

dcsiebe committed:
feat(feedback): change to feedback result

Rename feedback to feedback result as this seems to be more appropriate.
A param (resultSetId) is also added as the results are contained in a
set.

dcsiebe committed:
refactor(progress): use recording sets

Instead of having only a single recording per challenge, use a recording
set to allow multiple recordings per challenge.
This is because an 'answer' can be divided amongst multiple recordings.

kvdb reviewed (approved):
This reflects what we've discussed offline.

dcsiebe committed:
feat(progress): add audio url to result set

Add an audio url to the recording result set. This is the concatenated
audio of all recordings in the set.

dcsiebe committed:
feat(recordings): resultSetId param

Add the resultSetId to the recordings api and document some things about
the recording result set.

kvdb commented on docs/api/progress.md:
```diff
@@ -42,6 +42,7 @@ Content-Type: application/json
                 "audioUrl": "http://example.com/audio.mp3",
                 "recordingSet": {
                   "id": "5600160f578c4db6b7b0327e16e42153",
+                  "audioUrl": "http://example.com/fullaudio.mp3",
```
you don't want to call it fullAudioUrl? Guess it's obvious from the structure, right?

joggienl replied:
Ah, I made the same comment I see.

dcsiebe replied:
Imo it's the audio of a result set and that's what `audioUrl` describes, simple. I'd say that `fullAudioUrl` doesn't necessarily make more sense; it's the full audio - so other audio is partial, or empty?
If it should be more self describing I'd then opt for `concatenatedAudioUrl` but that is a bit lengthy, right?

joggienl replied:
just `audioUrl` is to vague. what about `combinedAudioUrl`? That would be a bit shorter.

dcsiebe replied:
Fine by me.

joggienl replied:
👍

joggienl commented on docs/api/recordings.md:
```diff
@@ -138,3 +139,18 @@ Content-Type: application/json
+of the result set.
+
+The result set cannot be queried directly but it can be seen when getting
+[progress](progress.md).
```
This implies that I would have to use the progress api to playback audio for the student. That is not what currently happens.

I was under the impression that the GET for a SpeechRecording would also return the the collection (=set) right?
So it would return either the the recording (just as it does now) or it returns the set of recordings?

dcsiebe replied:
The recordings api returns a recording (without the result set). The result set id should be in there (I forgot to add that).
This is why I opted for a separate result set api.

joggienl replied:
Maybe it is also a good idea to put the full audio url in there as well, that way I can still use the recording API as is...

dcsiebe replied:
:thinking: how? Embedding the result set partially or adding the param to the recording?

joggienl replied:
Something like this:

### Request

 ```http
 GET /challenges/speech/speech_1/recordings/recording_1 HTTP/1.1
 Accept: application/json
 ```
 
 ### Response
 
 ```http
 HTTP/1.1 200 OK
 Content-Type: application/json
 
 {
   "id": "speech_1",
   "created": "2014-01-28T21:25:10Z",
   "updated": "2014-01-28T21:25:10Z",
   "audioUrl": "https://api.itslanguage.nl/download/YsjdG37bUGseu8-bsJ",
   "userId": "24",
   "resultSetId": "4everAndever",
   "combinedAudioUrl": "https://api.itslanguage.nl/download/4everAndever"
 }
```

or as result:
```http
 HTTP/1.1 200 OK
 Content-Type: application/json
 
 {
   "id": "speech_1",
   "created": "2014-01-28T21:25:10Z",
   "updated": "2014-01-28T21:25:10Z",
   "audioUrl": "https://api.itslanguage.nl/download/YsjdG37bUGseu8-bsJ",
   "userId": "24",
   "resultSet": {
      "id": "4everAndever",
      "combinedAudioUrl": "https://api.itslanguage.nl/download/4everAndever"
   }
}
```

dcsiebe replied:
I think the embedded `resultSet` object makes more sense, simply adding the audio attribute to the recording implies that it belongs to the recording. Will make the change.

joggienl commented on docs/api/progress.md:
```diff
@@ -40,15 +40,25 @@ Content-Type: application/json
-                    "audioUrl": "http://example.com/audio.mp3"
+                "recordingSet": {
+                  "id": "5600160f578c4db6b7b0327e16e42153",
+                  "audioUrl": "http://example.com/fullaudio.mp3",
```
Maybe it is not clear (enought) that the audioUrl here is the concatenated audio for all the recordings toghether?

dcsiebe replied:
https://github.com/itslanguage/itslanguage-docs/pull/83#discussion_r213652761

dcsiebe committed:
fix(recordings): add resultSetId param to example

The example request didn't contain the result set id param, since it it
a required param let's have it in the example.

dcsiebe committed:
feat(recordings): embed result set

Instead of only returning the id of the result set, return an embedded
version of the result set. This includes the audio url so that audio can
be used directly for convenience.

dcsiebe committed:
refactor(progress): rename result set audio attr

Rename the audioUrl attr of the result set to combinedAudioUrl so it is
more self describing.
